### PR TITLE
pulse Schedule not filter

### LINKED ISSUE #3303: Add a not filter for schedules
There is a filter for schedules (https://github.com/Qiskit/qiskit-terra/blob/a3ae434326aefbb19a7861518b4a4f1e05528fb1/qiskit/pulse/schedule.py#L232) It would be convenient if that returned the remainder of the schedule that doesn't satisfy the filter. This makes it easier to, e.g., shift a part of a schedule. 

`filter(schedule, blah) << 100 + filter(schedule, not blah)`

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

Closes #3303, by adding an option `filter_type` to `Schedule.filter`

### CODE REVIEW

File: qiskit/pulse/schedule.py [hunk 1]
```diff
@@ -230,14 +231,30 @@ def filter(self, *filter_funcs: List[Callable],
                channels: Optional[Iterable[Channel]] = None,
                instruction_types: Optional[Iterable[Type['Instruction']]] = None,
                time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
-               intervals: Optional[Iterable[Interval]] = None) -> 'Schedule':
+               intervals: Optional[Iterable[Interval]] = None,
+               filter_type: Optional[int] = 0) -> Union['Schedule', Tuple['Schedule', 'Schedule']]:
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: Arbitrary integers are hard for the user to learn. The method name should tell the user what they can expect the method to do. It's also good for a function to do one thing reliably. Can you please make a new method that acts as a not filter? On the issue, I mentioned `exclude` as the method name, but other names could also work.

File: qiskit/pulse/schedule.py [hunk 2]
```diff
@@ -230,14 +231,30 @@ def filter(self, *filter_funcs: List[Callable],
...
+        be created. For intervals, a filter returning True on instructions wholly
+        contained within *any* of the given intervals will be created.
+
+        The behaviour types:
+
+        If filter_type == 0:
+            returns a Schedule containing instructions returning True on all filters.
+        If filter_type == 1:
+            returns a Schedule containing the instructions excluded by filter_type==0
+        If filter_type == 2:
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: Let's remove this option and only do "filter" and "not filter". As @dcmckayibm put in the issue, the goal is simply:
`schedule.filter(args) | schedule.exclude(args) == schedule`

File: test/python/pulse/test_schedule.py [hunk 3]
```diff
@@ -431,10 +431,16 @@ def test_delay_snapshot_channel(self):
 
 
 class TestScheduleFilter(BaseTestSchedule):
-    """Test Schedule filtering methods."""
-
+    """
+    Test Schedule filtering methods: Schedule.filter and Schedule.exclude
+
+    Note: each test in this class uses the helper: self.filter_and_test_consistency
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: It is not necessary to document this at the class level. This is an internal helper function, not really necessary for someone to understand this outright. We could even make `filter_and_test_consistency` a private method

Furthermore, comments are hard to maintain; nothing will ever break if a comment is outdated, but it may mislead developers. This documentation should therefore remain close to the function it describes. This improves the chances that we won't end up with this comment remaining in the future if things change, for instance if that helper function is removed.

File: test/python/pulse/test_schedule.py [hunk 4]
```diff
@@ -518,31 +564,77 @@ def test_filter_multiple(self):
...
+        self.assertEqual(len(excluded.instructions), 2)
+
+        # multiple custom filters specified as unkeyed arguments
+        filtered, excluded = self.filter_and_test_consistency(sched,
+                                                              lambda x: x[0] > 0,
+                                                              lambda x: x[0] < 30)
+        self.assertEqual(len(filtered.instructions), 1)
+        self.assertEqual(len(excluded.instructions), 2)
+
+    def filter_and_test_consistency(self, schedule, *args, **kwargs):
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: I think we are using type hints in these files

File: test/python/pulse/test_schedule.py [hunk 5]
```diff
@@ -518,31 +564,77 @@ def test_filter_multiple(self):
...
+        """
+        This is a helper function which returns the tuple
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: This should be written in a "command" or "recipe" form: `Return the tuple... and assert that ....`

File: test/python/pulse/test_schedule.py [hunk 6]
```diff
@@ -518,31 +564,77 @@ def test_filter_multiple(self):
...
+        """Test custom filters."""
         device = self.two_qubit_device
         lp0 = self.linear(duration=3, slope=0.2, intercept=0.1)
         sched = Schedule(name='fake_experiment')
         sched = sched.insert(0, lp0(device.drives[0]))
         sched = sched.insert(10, lp0(device.drives[1]))
         sched = sched.insert(30, FrameChange(phase=-1.57)(device.drives[0]))
-        for i in sched._filter([lambda x: True]).instructions:
+
+        filtered, excluded = self.filter_and_test_consistency(sched, [lambda x: True])
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: Ah, so, this should not be passed as a list. The `*` before `filter_funcs` splices the args for you. So, you'd write `self.filter...tency(sched, lambda x: True)` and then the variable `filter_funcs` will be assigned to `[lambda x: True]`.

Then you can remove those extra processing lines you had above.

File: test/python/pulse/test_schedule.py [hunk 7]
```diff
@@ -431,10 +431,16 @@ def test_delay_snapshot_channel(self):
...
+    Test Schedule filtering methods: Schedule.filter and Schedule.exclude
+
+    Note: each test in this class uses the helper: self.filter_and_test_consistency
+    the purpose of which is to return the outputs of both Schedule.filter and Schedule.exclude
+    in a tuple, while simultaneously testing that for a given Schedule sched,
+    sched.filter | sched.exclude == sched
+    """
     def test_filter_channels(self):
-        """Test filtering over channels."""
+        """Test filtering over channels"""
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: Revert this change

File: test/python/pulse/test_schedule.py [hunk 8]
```diff
@@ -463,16 +477,30 @@ def test_filter_inst_types(self):
         sched = sched.insert(60, acquire(device.acquires, device.memoryslots))
         sched = sched.insert(90, lp0(device.drives[0]))
 
-        only_acquires = sched.filter(instruction_types=[AcquireInstruction])
-        for _, inst in only_acquires.instructions:
+        # split into parts with and without AcquireInstruction
```
Comment (lcapelluto) [inline test/python/pulse/test_schedule.py]: The code that you've written here is self explanatory. (yay!) `only_acquire, no_acquire = self.filter(schedule, ...)` tells me that I am splitting the schedule into pieces with acquires and pieces without. When code is clear, we don't need to repeat the same info in a comment. Some people do love having superfluous comments, so this is just one way to think about it. Usually, when I write a comment, I ask myself, is there a way to rewrite my code that this comment wouldn't be necessary? Here, this test is doing a lot. We are using comments mainly as headers to say "test on Acquire", "now test of two other instruction types", "lastly just the framechange". We could shorten the comments to highlight that they _are_ headers and just write the instruction types we're testing, or we could make this into three separate tests.

File: qiskit/pulse/schedule.py [hunk 9]
```diff
@@ -249,6 +250,60 @@ def filter(self, *filter_funcs: List[Callable],
...
+                channels: Optional[Iterable[Channel]] = None,
+                instruction_types: Optional[Iterable[Type['Instruction']]] = None,
+                time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
+                intervals: Optional[Iterable[Interval]] = None) -> 'Schedule':
+        """
+        Return a Schedule with only the instructions *failing* at least one of the provided filters.
+        That is, this function has the same inputs as self.filter, but returns the complement
+        schedule, so that self.filter(args) | self.exclude(args) == self
+        """
+
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: bonus whitespace
```suggestion
```

File: qiskit/pulse/schedule.py [hunk 10]
```diff
@@ -249,6 +250,60 @@ def filter(self, *filter_funcs: List[Callable],
...
+    def exclude(self, *filter_funcs: List[Callable],
+                channels: Optional[Iterable[Channel]] = None,
+                instruction_types: Optional[Iterable[Type['Instruction']]] = None,
+                time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
+                intervals: Optional[Iterable[Interval]] = None) -> 'Schedule':
+        """
+        Return a Schedule with only the instructions *failing* at least one of the provided filters.
+        That is, this function has the same inputs as self.filter, but returns the complement
+        schedule, so that self.filter(args) | self.exclude(args) == self
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: Document args. The arg documentation for these three methods should all be the same

File: qiskit/pulse/schedule.py [hunk 11]
```diff
@@ -249,6 +250,60 @@ def filter(self, *filter_funcs: List[Callable],
             time_ranges: Time intervals to keep, e.g. [(0, 5), (6, 10)]
             intervals: Time intervals to keep, e.g. [Interval(0, 5), Interval(6, 10)]
         """
+        total_filter = self._construct_filter(*filter_funcs,
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: For some reason, I feel like `composed_filter` would have been more clear to me on first read. I don't feel strongly though, it's fine if you leave it as `total_filter`.

File: qiskit/pulse/schedule.py [hunk 12]
```diff
@@ -269,49 +324,32 @@ def interval_filter(time_inst: Tuple[int, 'Instruction']) -> bool:
                 return False
             return interval_filter
 
-        filter_funcs = list(filter_funcs)
+        # Below deals with the way custom filter_funcs might be specified; i.e. either in a list
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: I think I resolved this in a later comment on the test file

File: qiskit/pulse/schedule.py [hunk 13]
```diff
@@ -249,6 +250,60 @@ def filter(self, *filter_funcs: List[Callable],
...
+    def _construct_filter(self, *filter_funcs: List[Callable],
+                          channels: Optional[Iterable[Channel]] = None,
+                          instruction_types: Optional[Iterable[Type['Instruction']]] = None,
+                          time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
+                          intervals: Optional[Iterable[Interval]] = None) -> Callable:
+        """
+        Returns a boolean-valued function with input type (int, ScheduleComponent) that returns True
+        if the input satisfies all of the criteria specified by the arguments, described below.
+
+        Args:
```
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: oh, by main description I meant the main variable description. I feel like it would be easier to parse that way. What do you think?

[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/pulse/schedule.py]: I ended up deciding to keep that piece of the parameter description in the method description, as this time around I found it made it a bit more clear how each parameter influences the method behavior

File: qiskit/pulse/schedule.py [hunk 14]
```diff
@@ -42,6 +42,7 @@ def __init__(self, *schedules: List[Union[ScheduleComponent, Tuple[int, Schedule
         Raises:
             PulseError: If timeslots intercept.
         """
+
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]:
```suggestion
```

File: qiskit/pulse/schedule.py [hunk 15]
```diff
@@ -232,12 +233,11 @@ def filter(self, *filter_funcs: List[Callable],
                time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
                intervals: Optional[Iterable[Interval]] = None) -> 'Schedule':
         """
-        Return a new Schedule with only the instructions which pass though the provided filters.
-        Custom filters may be provided. If a list of channel indices is provided, only the
-        instructions that involve that channel (and maybe also others) will be included in the new
-        schedule. Similarly for instruction_types, only the instructions which are instances of the
-        provided types will be included. For intervals, instructions will be retained if their
-        timeslots are all wholly contained within *any* of the given intervals.
+        Return a new Schedule with only the instructions from this Schedule passing though the
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]:
```suggestion
        Return a new Schedule with only the instructions from this Schedule which pass though the
```

File: qiskit/pulse/schedule.py [hunk 16]
```diff
@@ -246,9 +246,78 @@ def filter(self, *filter_funcs: List[Callable],
...
+                                  new_sched_name="{name}-filtered".format(name=self.name))
+
+    def exclude(self, *filter_funcs: List[Callable],
+                channels: Optional[Iterable[Channel]] = None,
+                instruction_types: Optional[Iterable[Type['Instruction']]] = None,
+                time_ranges: Optional[Iterable[Tuple[int, int]]] = None,
+                intervals: Optional[Iterable[Interval]] = None) -> 'Schedule':
+        """
+        Return a Schedule with only the instructions from this Schedule *failing* at least one of
+        the provided filters. That is, this function has the same inputs as self.filter, but returns
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]:
```suggestion
        the provided filters. This method is the complement of `self.filter`, so that
            self.filter(args) | self.exclude(args) == self
```

File: qiskit/pulse/schedule.py [hunk 17]
```diff
@@ -246,9 +246,78 @@ def filter(self, *filter_funcs: List[Callable],
...
+        the complement Schedule, so that self.filter(args) | self.exclude(args) == self
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: Can't do multi-line suggestion
```suggestion
```

File: qiskit/pulse/schedule.py [hunk 18]
```diff
@@ -246,9 +246,78 @@ def filter(self, *filter_funcs: List[Callable],
...
+
+        Args:
+            filter_funcs: A list of Callables which take a (int, ScheduleComponent) tuple and
+                          return a bool
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]:
```suggestion
                          return a bool.
```

File: qiskit/pulse/schedule.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/schedule.py]: 👍

File: qiskit/pulse/schedule.py [hunk 19]
```diff
@@ -246,9 +246,78 @@ def filter(self, *filter_funcs: List[Callable],
...
+        return self._apply_filter(lambda x: not composed_filter(x),
+                                  new_sched_name="{name}-excluded".format(name=self.name))
+
+    def _apply_filter(self, filter_func: Callable, new_sched_name: str) -> 'Schedule':
+        """
+        Return a Schedule containing only the instructions from this Schedule that
+        filter_func returns True on
+
+        Args:
+            filter_func: function of the form (int, ScheduleComponent) -> bool
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: should end with periods I think

File: qiskit/pulse/schedule.py [hunk 20]
```diff
@@ -246,9 +246,78 @@ def filter(self, *filter_funcs: List[Callable],
...
+                                                 channels=channels,
+                                                 instruction_types=instruction_types,
+                                                 time_ranges=time_ranges,
+                                                 intervals=intervals)
+        return self._apply_filter(lambda x: not composed_filter(x),
+                                  new_sched_name="{name}-excluded".format(name=self.name))
+
+    def _apply_filter(self, filter_func: Callable, new_sched_name: str) -> 'Schedule':
+        """
+        Return a Schedule containing only the instructions from this Schedule that
```
Comment (lcapelluto) [inline qiskit/pulse/schedule.py]: Okay, once this docstring is written, should be good to go!
